RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1652 @@
+      }
+    }
+  };
+
+  // Process topo clusters using multi-seed iterative algorithm on serial CPU
+  class FastClusterSerial {
+  public:
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(const TAcc& acc,
+                                  const reco::PFRecHitHostCollection::ConstView pfRecHits,
+                                  const reco::PFClusterParamsDeviceCollection::ConstView pfClusParams,
+                                  reco::PFClusteringVarsDeviceCollection::View pfClusteringVars,
+                                  reco::PFClusterDeviceCollection::View clusterView,
+                                  reco::PFRecHitFractionDeviceCollection::View fracView) const {
+      const int nRH = pfRecHits.size();
+      int& topoId = alpaka::declareSharedVar<int, __COUNTER__>(acc);
+      int& nRHTopo = alpaka::declareSharedVar<int, __COUNTER__>(acc);
+      int& nSeeds = alpaka::declareSharedVar<int, __COUNTER__>(acc);
+
+      if (alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u] == 0) {
+        topoId = alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0u];
+        nRHTopo = pfClusteringVars[topoId].topoRHCount();
+        nSeeds = pfClusteringVars[topoId].topoSeedCount();
+      }
+
+      alpaka::syncBlockThreads(acc);  // all threads call sync
+
+      if (topoId < nRH && nRHTopo > 0 && nSeeds > 0) {
+        if (nRHTopo == nSeeds) {
+          // PF cluster is isolated seed. No iterations needed
+          if (alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u] == 0) {
+            // Fill PFCluster-level information
+            int rhIdx = pfClusteringVars[pfClusteringVars[topoId].topoSeedOffsets()]
+                            .topoSeedList();  // i is the seed rechit index
+            int seedIdx = pfClusteringVars[rhIdx].rhIdxToSeedIdx();
+            clusterView[seedIdx].energy() = pfRecHits[rhIdx].energy();
+            clusterView[seedIdx].x() = pfRecHits[rhIdx].x();
+            clusterView[seedIdx].y() = pfRecHits[rhIdx].y();
+            clusterView[seedIdx].z() = pfRecHits[rhIdx].z();
+          }
+        } else if (nSeeds <= 400 && nRHTopo - nSeeds <= 1500) {
+          dev_hcalFastCluster_multiSeedIterative(
+              acc, pfClusParams, topoId, nSeeds, nRHTopo, pfRecHits, pfClusteringVars, clusterView, fracView);
+        } else {
+          if (alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u] == 0)
+            printf("Topo cluster %d has %d seeds and %d rechits. Will be processed in next kernel.\n",
+                   topoId,
+                   nSeeds,
+                   nRHTopo);
+        }
+      } else
+        return;
+    }
+  };
+
+  PFClusterProducerKernel::PFClusterProducerKernel(Queue& queue, const reco::PFRecHitHostCollection& pfRecHits)
+      : nSeeds(cms::alpakatools::make_device_buffer<uint32_t>(queue)),
+        globalClusterPos(cms::alpakatools::make_device_buffer<Position4[]>(queue, 4 * 2304)),
```

Comment:
why `4 * 2304` ?
can you define a named constant for 2304 and use it everywhere ?

Reply:
Number of blocks * largest number of single depth HB channels. This channel number represents the worst case scenario for a topo cluster and will probably never happen, but the ability to process them is there.

For defining constants, is there a recommendation on where to put these? Is within the same file okay or is it better defined somewhere else?

Reply:
If it's only ever used in this file, defining it here it's OK.
If it's shared across multiple files, it should go to some reasonable header file.